File: 211. Add and Search Word - Data structure design.cpp
class WordDictionary {
public:
    struct trieNode {
        bool isWord;
        trieNode* children[26];
        trieNode(): isWord(false) {
            for (int i = 0; i < 26; i++) {
                children[i] = NULL;
            }
        }
    };

    trieNode *root;
    /** Initialize your data structure here. */
    WordDictionary() {
        root = new trieNode();
    }

    /** Adds a word into the data structure. */
    void addWord(string word) {
        trieNode *head = root;
        for (int i = 0; i < word.size(); i++) {
            int idx = word[i] - 'a';
            if (!head->children[idx]) {
                head->children[idx] = new trieNode();
            }
            head = head->children[idx];
        }
        head->isWord = true;
    }

    /** Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter. */
    bool search(string word) {
        deque<trieNode*> Q;
        Q.push_back(root);
        for (int i = 0; i < word.size(); i++) {
            int size = Q.size();
            if (word[i] >= 'a' && word[i] <= 'z') {
                int idx = word[i] - 'a';
                for (int j = 0; j < size; j++) {
                    trieNode* tp = Q.front();
                    Q.pop_front();
                    if (tp->children[idx]) {
                        Q.push_back(tp->children[idx]);
                    }
                }
            } else if (word[i] == '.') {
                for (int j = 0; j < size; j++) {
                    trieNode* tp = Q.front();
                    Q.pop_front();
                    for (int k = 0; k < 26; k++) {
                        if (tp->children[k]) {
                            Q.push_back(tp->children[k]);
                        }
                    }
                }
            }
            if (Q.empty()) return false;
        }

        for (auto it = Q.begin(); it != Q.end(); it++) {
            if ((*it)->isWord) {
                return true;
            }
        }
        return false;
    }
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * bool param_2 = obj.search(word);
 */

File: java/211. Add and Search Word - Data structure design.java
// O(n) trie
class WordDictionary {
    private class TrieNode {
        TrieNode[] letters;
        boolean isValid;
        TrieNode() {
            letters = new TrieNode[26];
            isValid = false;
        }
    }
    private TrieNode head;
    /** Initialize your data structure here. */
    public WordDictionary() {
        head = new TrieNode();
    }

    /** Adds a word into the data structure. */
    public void addWord(String word) {
        TrieNode cur = head;
        for (int i = 0; i < word.length(); ++i) {
            int n = word.charAt(i) - 'a';
            if (cur.letters[n] == null) {
                cur.letters[n] = new TrieNode();
            }
            cur = cur.letters[n];
        }
        cur.isValid = true;
    }

    /** Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter. */
    public boolean search(String word) {
        return dfs(head, word, 0);
    }

    private boolean dfs(TrieNode cur, String word, int idx) {
        if (idx == word.length()) {
            return cur.isValid;
        }
        if (word.charAt(idx) != '.') {
            if (cur.letters[word.charAt(idx) - 'a'] == null) {
                return false;
            }
            return dfs(cur.letters[word.charAt(idx) - 'a'], word, idx + 1);
        } else {
            for (int i = 0; i < 26; ++i) {
                if (cur.letters[i] != null && dfs(cur.letters[i], word, idx + 1)) {
                    return true;
                }
            }
        }
        return false;
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */
